greedy_algorithms/job_sequencing_algorithm.cpp

```diff
@@ -0,0 +1,197 @@
+ * ------|-----------|----------
+ * a     |      4    |    20
+ * b     |      1    |    10
+ * c     |      1    |    40
+ * d     |      1    |    30
+ *
+ * Output: Following is maximum profit sequence of jobs: c, a
+ *
+ *  Input:  Five Jobs with following deadlines and profits
+ *  JobID | Deadline  | Profit
+ * -------|-----------|--------
+ *   a    |     2     |  100
+ *   b    |     1     |  19
+ *   c    |     2     |  27
+ *   d    |     1     |  25
+ *   e    |     3     |  15
+ * Output: Following is maximum profit sequence of jobs: c, a, e
+ *
+ *
+ *
+ * Naive Approach:
+ * Generate all subsets of a given set of jobs and check individual subsets for
+ * the feasibility of jobs in that subset. Keep track of maximum profit among
+ * all feasible subsets.
+ * The above approach's time complexity is: O(N^2) with auxliary space O(N)
+ * N refers to the number of jobs.
+ *
+ *
+ * Greedy Approach:
+ * Greedily choose the jobs with maximum profit first, by sorting the jobs in
+ * decreasing order of their profit. This would help to maximize the total
+ * profit as choosing the job with maximum profit for every time slot will
+ * eventually maximize the total profit.
+ * The above approaches's time complexity is: O(NlogN) with auxliary space o(N)
+ * N refers to the number of jobs.
+ *
+ *
+ * Greedy approach psuedo-code:
+  - Sort all jobs in decreasing order of profit.
+  - Iterate on jobs in decreasing order of profit.
+       - For each job , do the following:
+           - Find a time slot i, such that slot is empty and i < deadline and i
+             is greatest.
+           - Put the job in this slot and mark this slot filled.
+           - If no such i exists, then ignore the job.
+ *
+ * @author [Mahmoud Elkholy](https://github.com/maahmoudezzat)
+ */
+
+// Implementation of the greedy algorithm using priority_queue AKA max_heap.
+// C++ code for the above approach
+#include <algorithm>  // for sort function.
+#include <cassert>    // for assert funtion in testing
+#include <iostream>   // for standard in/out
+#include <queue>      // for priority queue data structure
+#include <vector>     // for vector data structure
+
+/**
+ * @namespace
+ * @brief Greedy Algorithms
+ */
+namespace greedy_algorithms {
+
+/**
+ * @brief A structure to represent a job
+ */
+struct Job {
+    char id;     // Job Id
+    int dead;    // Deadline of job
+    int profit;  // Profit earned if job is completed before deadline
+};  // namespace greedy_algorithmsstruct Job
+
+// Custom sorting helper struct which is used for sorting
+// all jobs according to profit
+
+/**
+ * @brief Utility function that finds Custom sorting helper
+ * struct which is used for sorting all jobs according to profit
+ *
+ * @param a first job struct
+ * @param b second job struct
+ * @returns true if the proft of a less than profit of b, else false.
+ */
+struct jobProfit {
+    bool operator()(Job const& a, Job const& b) {
+        return (a.profit < b.profit);
```

Comment:
why not add this to the job struct?

Reply:
For future extension reasons, there are variations of this algorithm so it expected to add more utility functions to this struct.

Reply:
we will get to it when we get to it. since the algorithm currently only offers one utility function better merge with the implementation

Reply:
We could extract it out from the JobProfit struct but we couldn't add it to the Job struct since this comparison function should be defined as a static function or as a separate function outside the Job struct because C++ lambdas and standard sorting expect a static function or a functional object.

I have done an alternative solution for this case please have a look.